From 0e7883be1b0fcec3ae8d7aef78fdb76ba2f3990a Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Wed, 16 Aug 2023 15:17:58 +0800
Subject: [PATCH 01/52] mf: Add a session_flush_nodes() helper.

Signed-off-by: Zhiyi Zhang <zzhang@codeweavers.com>
(cherry picked from commit 8173d3b0ab3c620aaaa73b114b22826d2c4c20c4)

CW-Bug-Id: #22486
CW-Bug-Id: #22799
---
 dlls/mf/session.c | 37 +++++++++++++++++++++----------------
 1 file changed, 21 insertions(+), 16 deletions(-)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index f4806fb4c92..840d8150058 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -1053,6 +1053,26 @@ static void session_handle_start_error(struct media_session *session, HRESULT hr
     session_command_complete_with_event(session, MESessionStarted, hr, NULL);
 }
 
+static void session_set_state(struct media_session *session, enum session_state new_state)
+{
+    TRACE("setting state to %d\n", new_state);
+    session->state = new_state;
+    session_schedule_command(session);
+}
+
+static void session_flush_nodes(struct media_session *session)
+{
+    struct topo_node *node;
+
+    LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
+    {
+        if (node->type == MF_TOPOLOGY_OUTPUT_NODE)
+            IMFStreamSink_Flush(node->object.sink_stream);
+        else if (node->type == MF_TOPOLOGY_TRANSFORM_NODE)
+            IMFTransform_ProcessMessage(node->object.transform, MFT_MESSAGE_COMMAND_FLUSH, 0);
+    }
+}
+
 static void session_start(struct media_session *session, const GUID *time_format, const PROPVARIANT *start_position)
 {
     struct media_source *source;
@@ -3172,7 +3192,6 @@ static void session_set_source_object_state(struct media_session *session, IUnkn
     struct media_source *src;
     struct media_sink *sink;
     enum object_state state;
-    struct topo_node *node;
     BOOL changed = FALSE;
     DWORD i, count;
     HRESULT hr;
@@ -3268,21 +3287,7 @@ static void session_set_source_object_state(struct media_session *session, IUnkn
             if (!session_is_source_nodes_state(session, OBJ_STATE_STOPPED))
                 break;
 
-            LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
-            {
-                switch (node->type)
-                {
-                    case MF_TOPOLOGY_OUTPUT_NODE:
-                        IMFStreamSink_Flush(node->object.sink_stream);
-                        break;
-                    case MF_TOPOLOGY_TRANSFORM_NODE:
-                        IMFTransform_ProcessMessage(node->object.transform, MFT_MESSAGE_COMMAND_FLUSH, 0);
-                        break;
-                    default:
-                        ;
-                }
-            }
-
+            session_flush_nodes(session);
             session_set_caps(session, session->caps & ~MFSESSIONCAP_PAUSE);
 
             if (session->presentation.flags & SESSION_FLAG_FINALIZE_SINKS)
-- 
2.47.2

